components/eamxx/src/physics/mam/eamxx_mam_optics_process_interface.cpp

```diff
@@ -1,110 +1,627 @@
+
+    // it will syn data to device.
+    mam_coupling::read_water_refindex(table_name_water, grid_,
+                                      aerosol_optics_device_data_.crefwlw,
+                                      aerosol_optics_device_data_.crefwsw);
+    //
+    {
+      // make a list of host views
+      std::map<std::string, view_1d_host> host_views_aero;
+      // defines layouts
+      std::map<std::string, FieldLayout> layouts_aero;
+      ekat::ParameterList params_aero;
+      std::string surname_aero = "aer";
+      mam_coupling::set_refindex_names(surname_aero, params_aero, host_views_aero,
+                                 layouts_aero);
+
+      // read physical properties data for aerosol species
+      std::string mam4_soa_physical_properties_file =
+        m_params.get<std::string>("mam4_soa_physical_properties_file");
+
+      std::string mam4_dust_physical_properties_file =
+        m_params.get<std::string>("mam4_dust_physical_properties_file");
+
+      std::string mam4_nacl_physical_properties_file =
+        m_params.get<std::string>("mam4_nacl_physical_properties_file");
+
+      std::string mam4_so4_physical_properties_file =
+        m_params.get<std::string>("mam4_so4_physical_properties_file");
+
+      std::string mam4_pom_physical_properties_file =
+        m_params.get<std::string>("mam4_pom_physical_properties_file");
+
+      std::string mam4_bc_physical_properties_file =
+        m_params.get<std::string>("mam4_bc_physical_properties_file");
+
+      std::string mam4_mom_physical_properties_file =
+        m_params.get<std::string>("mam4_mom_physical_properties_file");
+
+      std::vector<std::string> name_table_aerosols = {
+        mam4_soa_physical_properties_file, mam4_dust_physical_properties_file,
+        mam4_nacl_physical_properties_file, mam4_so4_physical_properties_file,
+        mam4_pom_physical_properties_file, mam4_bc_physical_properties_file,
+        mam4_mom_physical_properties_file};
+
+      // specname_amode(ntot_aspectype) = (/ 'sulfate (0)   ',
+      //  'ammonium (1) ', 'nitrate (2)   ', &
+      //  'p-organic (3) ', 's-organic (4) ', 'black-c (5)  ', &
+      //  'seasalt (6)  ', 'dust  (7)    ', &
+      //  'm-organic (8)' /)
+      std::vector<int> species_ids = {
+          4,  // soa:s-organic
+          7,  // dst:dust:
+          6,  // ncl:seasalt
+          0,  // so4:sulfate
+          3,  // pom:p-organic
+          5,  // bc :black-c
+          8   // mom:m-organic
+      };
+
+      constexpr int maxd_aspectype = mam4::ndrop::maxd_aspectype;
+      auto specrefndxsw_host       = mam_coupling::complex_view_2d::HostMirror(
+                "specrefndxsw_host", nswbands_, maxd_aspectype);
+
+      auto specrefndxlw_host = mam_coupling::complex_view_2d::HostMirror(
+          "specrefndxlw_host", nlwbands_, maxd_aspectype);
+
+      const int n_spec = size(name_table_aerosols);
+      for(int ispec = 0; ispec < n_spec; ispec++) {
```

Comment:
Similar thing here. Why not
```c++
std::vector<std::string> specs = {"soa","dust","nacl",...};
for (const auto& spec : specs) {
   const auto table_name = "mam4_" + spec + "_physical_properties_file";
}

Reply:
You'd also have to change the species_ids to a map<string,int>, but that's also easier to read.

Reply:
Thanks again for these recommendations.